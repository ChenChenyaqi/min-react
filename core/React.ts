import { createElementVNode as createElement } from "./utils"
import type { Fiber, VNode } from "./types"
import { performWorkOfUnit } from "./fiberProcess"
import { commitDeletion, commitWork } from "./commit"

let nextWorkOfUnit: Fiber | null = null
// work in progress
let wipRoot: Fiber | null
// 当前处理的fiber
let wipFiber: Fiber | null
export const setWipFiber = (fiber: Fiber) => (wipFiber = fiber)
// 更新时需要
let currentRoot: Fiber | null
// 收集要删除的fiber
const deletions: Fiber[] = []

function workLoop(deadline: IdleDeadline) {
  let shouldYield = false
  while (!shouldYield && nextWorkOfUnit) {
    nextWorkOfUnit = performWorkOfUnit(nextWorkOfUnit, deletions)

    // 更新时，更新结束点是下一个兄弟节点之前
    if (nextWorkOfUnit?.type === wipRoot?.sibling?.type) {
      break
    }

    shouldYield = deadline.timeRemaining() < 1
  }
  if (!nextWorkOfUnit && wipRoot) {
    // 统一提交
    commitRoot()
  }
  if (!__TEST__) {
    requestIdleCallback(workLoop)
  }
}

function commitRoot() {
  deletions.forEach(commitDeletion)
  commitWork(wipRoot?.child)
  currentRoot = wipRoot
  wipRoot = null
  deletions.splice(0)
}

function render(el: VNode, container: HTMLElement) {
  wipRoot = {
    dom: container,
    props: {
      children: [el],
    },
  }
  nextWorkOfUnit = wipRoot
  requestIdleCallback(workLoop)
}

function update() {
  const currentFiber = wipFiber
  return () => {
    wipRoot = {
      ...currentFiber!,
      alternate: currentFiber, // 绑定老的fiber
    }
    nextWorkOfUnit = wipRoot
    requestIdleCallback(workLoop)
  }
}

const React = {
  render,
  createElement,
  update,
}

export default React
