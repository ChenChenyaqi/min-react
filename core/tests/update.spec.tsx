import React from "@core/React"
import { it, expect, describe, beforeEach, vi } from "vitest"

describe("core/React/update", () => {
  describe("manual update", () => {
    let count = 0
    const countCalled = vi.fn()
    let countUpdate
    const setCount = () => {
      count++
      countUpdate()
    }
    function Counter() {
      countCalled()
      countUpdate = React.update()
      return <div>{count}</div>
    }

    let foo = 0
    const fooCalled = vi.fn()
    let fooUpdate
    const setFoo = () => {
      foo++
      fooUpdate()
    }
    function Foo() {
      fooCalled()
      fooUpdate = React.update()
      return <div>{foo}</div>
    }
    beforeEach(() => {
      count = 0
      foo = 0
      countCalled.mockClear()
      fooCalled.mockClear()
    })
    it("should update function component when called React.update", () => {
      const root = document.createElement("div")
      function App() {
        return <Counter></Counter>
      }
      React.render(<App />, root)
      setCount()
      expect(count).toBe(1)
      expect(root).toMatchSnapshot()
    })

    it("should update specific function component when called React.update", () => {
      const root = document.createElement("div")
      function App() {
        return (
          <div>
            <Counter></Counter>
            <Foo></Foo>
          </div>
        )
      }
      React.render(<App />, root)
      countCalled.mockClear()
      fooCalled.mockClear()
      setCount()
      expect(count).toBe(1)
      expect(fooCalled).toBeCalledTimes(0)
      expect(countCalled).toBeCalledTimes(1)
    })
  })
})
