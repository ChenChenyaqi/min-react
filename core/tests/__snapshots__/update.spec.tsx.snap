// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`core/React/update > manual update > should update function component when called React.update 1`] = `
<div>
  <div>
    1
  </div>
</div>
`;
